### DISCUSSION TITLE
Why new crate?

### DISCUSSION THREAD

**Discussion by doriath** [jcornaz/beancount_parser_2#28] [General]
If this is a rewrite of beancount-parser (and you also own it), why create a new crate instead of releasing the new code under v2 version?

My personal opinions:

1. `beancount-parser` imho is nicer name than `beancount_parser_2`
2. Less crates related to beancount parsing means less confusion (less chance that someone will just use old `beancount-parser` instead of the newer version)
3. You could still keep `v1` branch to keep the old version in case someone wanted to contribute some improvement to the old version.

**Comment by jcornaz** [comment]
This is a very good question, thanks for asking. 

First of all, to me, this is what it is. This new parser is not a new version of `beancount-parser`, It is a different, brand new beancount parser that was made without re-using *anything* from the old one.  To me creating a new crate feels like the right thing to do because this is how I perceive it: *A new thing*.

Second, I was not enthusiastic about publishing a breaking change. You may think it doesn't matter because this new crate is like a breaking change for anyone coming from `beancount-parser`. But there are some differences:

* The two crates can easily and clearly be used together at the same time. In fact, they can even (potentially) feature interoperability with each other. I actually plan to add an optional `v2` module in `beancount-parser` that would be a re-export of the `beancount_parser_2` crate. Very few rust developers know how to depend on 2 different major versions of the same crate at the same time. 
* By being a new crate instead of a major version, it doesn't create pressure on all users of `beancount-parser`, which may have felt forced to update ASAP otherwise. If `beancount-parser` works for someone, then it'll continue to work for as long as they want to use it.
* If it turns out the first crate was much better for some kind of scenario, then the two crates may diverge and have an independent life (in which case, I'd probably transfer ownership of the former crate to a new maintainer)

> `beancount-parser` imho is nicer name than `beancount_parser_2`

I agree, `beancount-parser 2.0.0` may be a nicer name than `beancount_parser_2 1.0.0`. But not by much. If seeing `beancount_parser_2` in your rust code really bothers you, you can write `beancount-parser = { package = "beancount_parser_2", version = "..." }` in your `Cargo.toml`. 

Maybe we can solve that naming problem more elegantly by finding a new fancy name? I don't have many ideas yet, but I am open to suggestions.

> Less crates related to beancount parsing means less confusion

Yes, of course. Anyway, they are both out now.

And I don't think it is that bad. Eventually, the sorting algorithms of https://lib.rs and https://crates.io will show `beancount_parser_2` first in any search about `beancount`  (probably after `beancount_parser_2` is stabilized)

We may try to clarify things better, notably by changing the crates description so that it jumps even more to the eyes in this kind of search: https://lib.rs/search?q=beancount

> less chance that someone will just use the old `beancount-parser` instead of the newer version

Would that be bad? If they use `beancount-parser` and it works for them why not? Maybe they even prefer it? I don't mind if they use it. I hope they'll have fun projects with it. If one cares about the maintenance state of their dependencies, then it is up to them to read the readme to see whether or not there is a discontinuation notice. But they do what they want, I don't mind.

---

I hope this helps understand why I created a new crate instead of publishing a new major version of `beancount-parser`.  I definitely understand the arguments in favor of a new major version of `beancount-parser` as I did seriously consider that option before deciding to create `beancount_parser_2`. And nothing is ever set in stone. If enough users/contributors really strongly prefer `beancount-parser 2.0.0` over `beancount_parser_2` I may consider it. But in my eyes, a new major version of `beancount-parser` is not significantly better than a new crate (if better at all).

**Comment by doriath** [comment]
Thank you very much for a detailed explanation, this fully explains why separate crate makes sense in this case.

Also, thank you for creating this crate and amazing support, quickly responding to issues and accepting PRs!

**Reply by jcornaz** [reply]
Thanks to your for your interest and contributions! :smiley:

**Comment by jcornaz** [comment]
Actually... I finally changed my mind and just published [`beancount-parser 2.0.0-beta.1`](https://github.com/jcornaz/beancount-parser/releases/tag/v2.0.0-beta.1)

I realized my reasons were not that strong. Yes, it did feel right to create `beancount_parser_2` at the beginning of the rewrite, because that is how I was thinking about it: *a new thing*. But now that the rewrite is kinda complete (feature-wise), it starts to feel strange to have a new crate instead of re-using the old one.

Even if I still think that `beancount-parser 2.0.0` is not significantly better `beancount_parser_2 1.0.0`, I eventually accepted that `beancount_parser_2 1.0.0` is not significantly better than `beancount-parser 2.0.0` either. And `beancount-parser` *is* a better name for the crate than `beancount_parser_2`.

I am now discontinuing `beancount_parser_2` in favor of `beancount-parser v2`

Sorry for the confusing back and forth. And thank you again for questioning that decision.
